File: examples/runnable/08-aws-containers/go/import_pipeline.go
// Package main holds the import pipeline (validate -> transform -> load) as a
// Workflow + Activities. The Go equivalent of the Java ImportWorkflow lab.
// Activities pass S3 *URIs* forward, never file bytes, so Workflow history stays
// small. The bodies fake the work; in the lab they'd call LocalStack S3.
package main

import (
	"context"
	"fmt"
	"hash/fnv"
	"strings"
	"time"

	"go.temporal.io/sdk/activity"
	"go.temporal.io/sdk/temporal"
	"go.temporal.io/sdk/workflow"
)

const TaskQueue = "transform"

// Validate rewrites the incoming URI to a /validated/ key.
func Validate(ctx context.Context, inputS3URI string) (string, error) {
	time.Sleep(250 * time.Millisecond)
	activity.RecordHeartbeat(ctx, "validated")
	return strings.Replace(inputS3URI, "/incoming/", "/validated/", 1), nil
}

// Transform rewrites the validated URI to a /transformed/ key.
func Transform(ctx context.Context, validatedS3URI string) (string, error) {
	time.Sleep(time.Second)
	activity.RecordHeartbeat(ctx, "transformed")
	return strings.Replace(validatedS3URI, "/validated/", "/transformed/", 1), nil
}

// Load returns a synthetic row count for the transformed object.
func Load(ctx context.Context, transformedS3URI string) (int64, error) {
	time.Sleep(500 * time.Millisecond)
	h := fnv.New32a()
	_, _ = h.Write([]byte(transformedS3URI))
	return int64(h.Sum32() % 10_000), nil
}

// ImportWorkflow runs validate -> transform -> load and returns the final URI.
func ImportWorkflow(ctx workflow.Context, inputS3URI string) (string, error) {
	ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
		StartToCloseTimeout: 2 * time.Minute,
		RetryPolicy: &temporal.RetryPolicy{
			InitialInterval: time.Second,
			MaximumAttempts: 3,
		},
	})

	var validatedURI string
	if err := workflow.ExecuteActivity(ctx, Validate, inputS3URI).Get(ctx, &validatedURI); err != nil {
		return "", err
	}
	var transformedURI string
	if err := workflow.ExecuteActivity(ctx, Transform, validatedURI).Get(ctx, &transformedURI); err != nil {
		return "", err
	}
	var rowCount int64
	if err := workflow.ExecuteActivity(ctx, Load, transformedURI).Get(ctx, &rowCount); err != nil {
		return "", err
	}
	return fmt.Sprintf("%s?rows=%d", transformedURI, rowCount), nil
}

File: examples/runnable/08-aws-containers/python/import_pipeline.py
"""The import pipeline (validate -> transform -> load) as a Workflow + Activities.

The Python equivalent of the Java ImportWorkflow lab. Activities pass S3 *URIs*
forward, never file bytes, so Workflow history stays small. The bodies here fake
the work with sleeps + URI rewriting; in the lab they'd call LocalStack S3.
"""

import asyncio
from datetime import timedelta

from temporalio import activity, workflow
from temporalio.common import RetryPolicy

TASK_QUEUE = "transform"


@activity.defn
async def validate(input_s3_uri: str) -> str:
    await asyncio.sleep(0.25)
    activity.heartbeat("validated")
    return input_s3_uri.replace("/incoming/", "/validated/")


@activity.defn
async def transform(validated_s3_uri: str) -> str:
    await asyncio.sleep(1)
    activity.heartbeat("transformed")
    return validated_s3_uri.replace("/validated/", "/transformed/")


@activity.defn
async def load(transformed_s3_uri: str) -> int:
    await asyncio.sleep(0.5)
    return abs(hash(transformed_s3_uri)) % 10_000


@workflow.defn
class ImportWorkflow:
    @workflow.run
    async def run(self, input_s3_uri: str) -> str:
        opts = dict(
            start_to_close_timeout=timedelta(minutes=2),
            retry_policy=RetryPolicy(
                initial_interval=timedelta(seconds=1), maximum_attempts=3
            ),
        )
        validated_uri = await workflow.execute_activity(validate, input_s3_uri, **opts)
        transformed_uri = await workflow.execute_activity(transform, validated_uri, **opts)
        row_count = await workflow.execute_activity(load, transformed_uri, **opts)
        return f"{transformed_uri}?rows={row_count}"
